doc/tutorials/interactive/codelabs/elektra-specification.md

```diff
@@ -0,0 +1,464 @@
+summary: How to Write a specification in Elektra
+id: elektra-specification
+categories: Elektra
+tags: cm2021
+status: Published 
+authors: @aaronabebe
+Feedback Link: https://libelektra.org
+
+# How to Write a specification in Elektra<!-- ------------------------ -->
+## Overview 
+Duration: 1
+
+![elektra-logo](assets/elektra-logo.svg)
+
+### Introduction
+In this codelab you will learn how to interactively use the `SpecElektra` specification language and `kdb` to write a configuration specification for an example application.
+
+### What you should already know
+- how to install elektra
+- very basic elektra commands and concepts (kdb get, kdb set, kdb ls)
+- how to open and use a terminal
+
+### What you’ll Learn 
+- how to create and mount a specification using `kdb`
+- how to add keys with different types, defaults and examples to your specification and how to validate them
+- the benefits of using `kdb` to generate a specification, instead of writing one by hand 
+
+### What you'll do
+- use code generation and `kdb` to create and mount a specification for an example CRUD application 
+- define defaults, examples and checks for keys in the validation 
+
+<!-- ------------------------ -->
+## Example App Overview
+Duration: 1
+
+For this tutorial we will write a specification for a very simple CRUD backend application. 
+You need to configure a `port` for the REST server, an `ip` and a SQL `dialect` for the database server the app 
+will connect to and finally a `date` where all the data will be saved to a backup.  
+
+So the application will need the following configuration options:
+- a server port
+- a database ip 
+- a database dialect
+- a backup date
+
+
+<!-- ------------------------ -->
+## Getting Started
+Duration: 3
+
+Make sure you have `Elektra` installed on your local machine. 
+```
+kdb --version
+
+KDB_VERSION: 0.9.5
+SO_VERSION: 5
+```
+
+Otherwise refer to the [getting started guide](https://www.libelektra.org/getstarted/guide) to install it.
+
+
+<!-- ------------------------ -->
+## Task: Mounting the specification
+Duration: 5
+
+### Step 1: Mount a specification file
+
+First you need to mount a specification file, in this case `spec.ni` to the `spec:/` namespace.
+We define the path inside the `spec:/` namespace as `/sw/app/current/#0`, this is an arbitrary name and you can choose whatever you want here.
+
+We also need the specify the plugin we use for writing to the file in the correct format. In this case we choose the `ni` plugin to write to the specification file.
+
+``` bash
+kdb mount ~/spec.ni spec:/sw/app/current/#0 ni
```

Comment:
```suggestion
sudo kdb mount `pwd`/spec.ni spec:/sw/app/current/#0 ni
```

mount should use sudo

Reply:
Neither `~/spec.ni` nor <code>&#96;pwd&#96;/spec.ni</code> will work as expected. Because you didn't specify a resolver, the default resolver will be used. In a standard setup this means the path will be seen as relative to `/usr/share/elektra/specification/`.

You should use one of these (the first one is the preferred solution)
- `sudo kdb mount spec.ni spec:/sw/app/current/#0 ni` (possibly with a more unique name instead of `spec.ni`)
- `sudo kdb mount --resolver=noresolver "$PWD/spec.ni" spec:/sw/app/current/#0 ni`

Reply:
Absolute paths for `spec:` and `system:` will stay absolute and work as expected. (Only for `user:` and `dir:` even absolute paths are not absolute.) As the shell resolves `~` it would be fine to use but I would not recommend to have specs in the home dir. With `pwd` it could be any path. Thus my suggestion.

> sudo kdb mount spec.ni spec:/sw/app/current/#0 ni (possibly with a more unique name instead of spec.ni)

Yes, that would also be okay. But then a person might be surprised where this spec.ni is. (`kdb file` would reveal the answer, though.)

> sudo kdb mount --resolver=noresolver "$PWD/spec.ni" spec:/sw/app/current/#0 ni

I would not recommend "noresolver", it will make storage plugins directly write to config files, which is not safe. "noresolver" was actually designed for non-file based plugins like uname. I created #3823 to clarify that.